cpp/locality/Voronoi.cc

```diff
@@ -0,0 +1,195 @@
+        if (n1.index_i != n2.index_i) {
+            return n1.index_i < n2.index_i;
+        }
+        if (n1.index_j != n2.index_j) {
+            return n1.index_j < n2.index_j;
+        }
+        return n1.weight < n2.weight;
+    }
+};
+
+typedef tbb::enumerable_thread_specific< std::vector<NeighborBond> > BondVector;
+
+void add_valid_bonds(BondVector::reference local_bonds,
+    unsigned int i, unsigned int expanded_i,
+    unsigned int j, unsigned int expanded_j,
+    unsigned int N, float weight)
+{
+    // Make sure we only add bonds with real particles as the reference
+    if (i < N)
+    {
+        NeighborBond nb_ij(expanded_i, expanded_j, weight);
+        local_bonds.emplace_back(nb_ij);
+    }
+
+    if (j < N)
+    {
+        NeighborBond nb_ji(expanded_j, expanded_i, weight);
+        local_bonds.emplace_back(nb_ji);
+    }
+}
+
+void Voronoi::compute(const box::Box &box, const vec3<double>* vertices,
+    const int* ridge_points, const int* ridge_vertices, unsigned int n_ridges,
+    unsigned int N, const int* expanded_ids, const int* ridge_vertex_indices)
+    {
+        m_box = box;
+
+        // iterate over ridges in parallel
+        BondVector bonds;
+        tbb::parallel_for(tbb::blocked_range<size_t>(0, n_ridges), [&] (const tbb::blocked_range<size_t> &r) {
+             BondVector::reference local_bonds(bonds.local());
+            for (size_t ridge(r.begin()); ridge != r.end(); ++ridge) {
+                unsigned int i = ridge_points[2*ridge];
+                unsigned int j = ridge_points[2*ridge+1];
+                float weight = 0;
+
+                // Reject bonds between two image particles
+                if (i >= N && j >= N)
+                    continue;
+
+                bool exclude_ii = true;
+                // We DO allow bonds from a particle to its own image
+                if (exclude_ii && i == j)
+                    continue;
+
+                // Bonds where a ridge goes to infinity (index -1) have weight 0
+                bool weighted_bond = true;
+                vector<int> current_ridge_vertex_ids;
+                for (int ridge_vert_id = ridge_vertex_indices[ridge]; ridge_vert_id < ridge_vertex_indices[ridge+1]; ++ridge_vert_id) {
+                    if (ridge_vertices[ridge_vert_id] == -1) {
+                        add_valid_bonds(local_bonds, i, expanded_ids[i], j, expanded_ids[j], N, 0);
+                        weighted_bond = false;
+                        break;
+                    } else {
+                        current_ridge_vertex_ids.push_back(ridge_vertices[ridge_vert_id]);
+                    }
+                }
+
+                if (weighted_bond) {
+                    if (box.is2D()) {
+                        // 2D weight is the length of the ridge edge
+                        int v1ind = current_ridge_vertex_ids[0];
+                        int v2ind = current_ridge_vertex_ids[1];
+                        vec3<double> v1 = vertices[v1ind];
```

Comment:
Can you add a comment near the top of this file indicating that you have to use double precision to get meaningful results? Otherwise this just looks wrong in freud, and it's also not obvious why in some cases you have double and others you have float.